projects/charger/src/charger_controller.c

```diff
@@ -0,0 +1,127 @@
+#include "charger_controller.h"
+
+#include <assert.h>
+#include <stdbool.h>
+#include <stdint.h>
+
+#include "can_interval.h"
+#include "can_transmit.h"
+#include "charger_events.h"
+#include "event_queue.h"
+#include "generic_can.h"
+#include "generic_can_msg.h"
+#include "status.h"
+
+#define CHARGER_PERIOD_US 1000000  // 1 Second as defined in datasheet.
+
+#define CHARGER_EXPECTED_RX_DLC 5
+#define CHARGER_EXPECTED_RX_ID 0x18FF50E5
+
+#define CHARGER_EXPECTED_TX_DLC 5
+#define CHARGER_EXPECTED_TX_ID 0x1806E5F4
+
+static CanInterval *s_interval;
+static ChargerStatus *s_charger_status;
+static GenericCanMsg s_tx_msg;
+
+typedef union J1939CanId {
+  uint32_t raw_id;
+  struct {
+    uint32_t source_address : 8;  // Source
+    uint32_t pdu_specifics : 8;   // Destination
+    uint32_t pdu_format : 8;      // Packet Format
+    uint32_t dp : 1;              // Always 0
+    uint32_t r : 1;               // Always 0
+    uint32_t priority : 3;        // Anything
+  };
+} J1939CanId;
+
+// Explicit for readability.
+static const J1939CanId s_rx_id = {
+  .source_address = 0xE5,  // Indicates BMS in the J1939 standard.
+  .pdu_specifics = 0x50,   // Broadcast address (BCA) id.
+  .pdu_format = 0xFF,      // From datasheet.
+  .dp = 0,
+  .r = 0,
+  .priority = 0x06,  // From datasheet.
+};
+static const J1939CanId s_tx_id = {
+  .source_address = 0xF4,  // Indicates BMS in the J1939 standard.
+  .pdu_specifics = 0xE5,   // Charger control system (CCS) id.
+  .pdu_format = 0x06,      // From datasheet.
+  .dp = 0,
+  .r = 0,
+  .priority = 0x06,  // From datasheet.
+};
+
+// GenericCanRx
+static void prv_rx_handler(const GenericCanMsg *msg, void *context) {
+  (void)context;
+  // TODO(ELEC-355): Rebroadcast this message after transform to the car.
+
+  const ChargerControllerRxData data = {
+    .raw_data = msg->data,
+  };
+  *s_charger_status = data.data_impl.status_flags;
+
+  // Check for statuses
+  if (!charger_controller_is_safe(data.data_impl.status_flags)) {
+    // If unsafe immediately stop charging and force the FSM to transition to the unsafe state.
+    charger_controller_set_state(CHARGER_STATE_STOP);
+    event_raise(CHARGER_EVENT_STOP_CHARGING, 0);
+  }
+}
+
+StatusCode charger_controller_init(ChargerSettings *settings, ChargerStatus *status) {
+  assert(s_tx_id.raw_id == CHARGER_EXPECTED_TX_ID);
```

Comment:
Isn't this something that could be handled in your test? (I think assert is disabled in release mode)

Reply:
Done.